File: tests/pilot_upload_cohort.spec.ts
import { test, expect, Page } from "@playwright/test";
import { signInTestUser } from "./shared";

let p: Page;

test("Pilot - upload cohort", async ({ page }) => {
  p = page;

  await given_the_app_is_setup();
  await and_i_am_signed_in();

  await when_i_visit_the_pilot_dashboard();
  await then_i_should_see_the_manage_pilot_page();

  await when_i_click_the_upload_cohort_link();
  await then_i_should_see_the_upload_cohort_page();

  await when_i_continue_without_uploading_a_file();
  await then_i_should_see_an_error();

  await when_i_upload_a_malformed_csv();
  await then_i_should_see_an_error();

  await when_i_upload_a_cohort_file_with_invalid_headers();
  await then_i_should_the_errors_page_with_invalid_headers();
  await and_i_should_be_able_to_go_back_to_the_upload_page();

  await when_i_upload_a_cohort_file_with_invalid_fields();
  await then_i_should_the_errors_page_with_invalid_fields();
  await and_i_should_be_able_to_go_to_the_upload_page();

  await when_i_upload_the_cohort_file();
  await then_i_should_see_the_success_page();
});

async function given_the_app_is_setup() {
  await p.goto("/reset");
}

async function and_i_am_signed_in() {
  await signInTestUser(p);
}

async function when_i_visit_the_pilot_dashboard() {
  await p.goto("/pilot");
}

async function then_i_should_see_the_manage_pilot_page() {
  await expect(
    p.getByRole("heading", { level: 1, name: "Manage pilot" }),
  ).toBeVisible();
}

async function when_i_click_the_upload_cohort_link() {
  await p.getByRole("link", { name: "Upload the cohort list" }).click();
}

async function then_i_should_see_the_upload_cohort_page() {
  await expect(
    p.getByRole("heading", { name: "Upload the cohort list" }),
  ).toBeVisible();
  await expect(p.locator(".nhsuk-inset-text")).toContainText(
    "Your current cohort has 21 children",
  );
}

async function when_i_upload_the_cohort_file() {
  await p.setInputFiles(
    'input[type="file"]',
    "spec/fixtures/cohort_list/valid_cohort.csv",
  );
  await p.getByRole("button", { name: "Upload the cohort list" }).click();
}

async function then_i_should_see_the_success_page() {
  await expect(
    p.getByRole("heading", { name: "Cohort data uploaded" }),
  ).toBeVisible();
}

async function when_i_continue_without_uploading_a_file() {
  await p.getByRole("button", { name: "Upload the cohort list" }).click();
}

async function then_i_should_see_an_error() {
  await expect(
    p.getByRole("heading", { name: "There is a problem" }),
  ).toBeVisible();
}

async function when_i_upload_a_malformed_csv() {
  await p.setInputFiles(
    'input[type="file"]',
    "spec/fixtures/cohort_list/malformed.csv",
  );
  await p.getByRole("button", { name: "Upload the cohort list" }).click();
}

async function when_i_upload_a_cohort_file_with_invalid_headers() {
  await p.setInputFiles(
    'input[type="file"]',
    "spec/fixtures/cohort_list/invalid_headers.csv",
  );
  await p.getByRole("button", { name: "Upload the cohort list" }).click();
}

async function then_i_should_the_errors_page_with_invalid_headers() {
  await expect(
    p.getByRole("heading", {
      level: 1,
      name: "The cohort list could not be added",
    }),
  ).toBeVisible();
  await expect(p.getByRole("heading", { level: 2, name: "CSV" })).toBeVisible();
}

async function and_i_should_be_able_to_go_back_to_the_upload_page() {
  await p.getByRole("link", { name: "Back to cohort upload page" }).click();
}

async function when_i_upload_a_cohort_file_with_invalid_fields() {
  await p.setInputFiles(
    'input[type="file"]',
    "spec/fixtures/cohort_list/invalid_fields.csv",
  );
  await p.getByRole("button", { name: "Upload the cohort list" }).click();
}

async function then_i_should_the_errors_page_with_invalid_fields() {
  await expect(
    p.getByRole("heading", {
      level: 1,
      name: "The cohort list could not be added",
    }),
  ).toBeVisible();
  await expect(
    p.getByRole("heading", { level: 2, name: "Row 2", exact: true }),
  ).toBeVisible();
}

async function and_i_should_be_able_to_go_to_the_upload_page() {
  await p.getByRole("link", { name: "Upload a new cohort list" }).click();
}

File: spec/features/pilot_upload_cohort_spec.rb
require "rails_helper"

RSpec.describe "Pilot - upload cohort" do
  scenario "User uploads a cohort list" do
    given_the_app_is_setup
    when_i_visit_the_pilot_dashboard
    then_i_should_see_the_manage_pilot_page

    when_i_click_the_upload_cohort_link
    then_i_should_see_the_upload_cohort_page

    when_i_continue_without_uploading_a_file
    then_i_should_see_an_error

    when_i_upload_a_malformed_csv
    then_i_should_see_an_error

    when_i_upload_a_cohort_file_with_invalid_headers
    then_i_should_the_errors_page_with_invalid_headers
    and_i_should_be_able_to_go_back_to_the_upload_page

    when_i_upload_a_cohort_file_with_invalid_fields
    then_i_should_the_errors_page_with_invalid_fields
    and_i_should_be_able_to_go_to_the_upload_page

    when_i_upload_the_cohort_file
    then_i_should_see_the_success_page
  end

  def given_the_app_is_setup
    @team = create(:team, :with_one_nurse)
    create(:location, team: @team, id: 1)
    @user = @team.users.first
  end

  def when_i_visit_the_pilot_dashboard
    sign_in @user
    visit "/pilot"
  end

  def then_i_should_see_the_manage_pilot_page
    expect(page).to have_content("Manage pilot")
  end

  def when_i_click_the_upload_cohort_link
    click_on "Upload the cohort list"
  end

  def then_i_should_see_the_upload_cohort_page
    expect(page).to have_content("Upload the cohort list")
  end

  def when_i_upload_the_cohort_file
    attach_file(
      "cohort_list[csv]",
      "spec/fixtures/cohort_list/valid_cohort.csv"
    )
    click_on "Upload the cohort list"
  end

  def then_i_should_see_the_success_page
    expect(page).to have_content("Cohort data uploaded")
  end

  def when_i_continue_without_uploading_a_file
    click_on "Upload the cohort list"
  end

  def then_i_should_see_an_error
    expect(page).to have_content("There is a problem")
  end

  def when_i_upload_a_malformed_csv
    attach_file("cohort_list[csv]", "spec/fixtures/cohort_list/malformed.csv")
    click_on "Upload the cohort list"
  end

  def when_i_upload_a_cohort_file_with_invalid_headers
    attach_file(
      "cohort_list[csv]",
      "spec/fixtures/cohort_list/invalid_headers.csv"
    )
    click_on "Upload the cohort list"
  end

  def then_i_should_the_errors_page_with_invalid_headers
    expect(page).to have_content("The cohort list could not be added")
    expect(page).to have_content("CSV")
  end

  def and_i_should_be_able_to_go_back_to_the_upload_page
    click_on "Back to cohort upload page"
  end

  def when_i_upload_a_cohort_file_with_invalid_fields
    attach_file(
      "cohort_list[csv]",
      "spec/fixtures/cohort_list/invalid_fields.csv"
    )
    click_on "Upload the cohort list"
  end

  def then_i_should_the_errors_page_with_invalid_fields
    expect(page).to have_content("The cohort list could not be added")
    expect(page).to have_content("Row 2")
  end

  def and_i_should_be_able_to_go_to_the_upload_page
    click_on "Upload a new cohort list"
  end
end
